ui/components/app/qr-hardware-state-watcher/base-reader.js

```diff
@@ -0,0 +1,210 @@
+      if (e.name === 'NotAllowedError') {
+        log.info(`Permission denied: '${e}'`);
+        setReady(READY_STATE.NEED_TO_ALLOW_ACCESS);
+      } else {
+        setError(e);
+      }
+    }
+  };
+
+  useEffect(() => {
+    mounted.current = true;
+    checkEnvironment();
+    return () => {
+      mounted.current = false;
+      clearTimeout(permissionChecker);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useEffect(() => {
+    if (ready === READY_STATE.READY) {
+      initCamera();
+    } else if (ready === READY_STATE.NEED_TO_ALLOW_ACCESS) {
+      checkPermissions();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [ready]);
+
+  const stopAndClose = () => {
+    history.goBack();
+  };
+
+  const tryAgain = () => {
+    clearTimeout(permissionChecker);
+    reset();
+    checkEnvironment();
+  };
+
+  const handleError = (e) => {
+    setError(e);
+  };
+
+  const handleScan = (data) => {
+    try {
+      if (!data) {
+        return;
+      }
+      urDecoder.receivePart(data);
+      if (urDecoder.isComplete()) {
+        const result = urDecoder.resultUR();
+        handleSuccess(result)
+          .then(() => stopAndClose())
+          .catch(setError);
+      }
+    } catch (e) {
+      setError(e);
+    }
+  };
+
+  const renderError = () => {
+    let title, msg;
+    if (error.type === 'NO_WEBCAM_FOUND') {
+      title = t('noWebcamFoundTitle');
+      msg = t('noWebcamFound');
+    } else if (error.message === t('unknownQrCode')) {
+      msg = t('unknownQrCode');
+    } else {
+      title = t('unknownCameraErrorTitle');
+      msg = t('unknownCameraError');
+    }
+
+    return (
+      <>
+        <div className="qr-scanner__image">
+          <img src="images/webcam.svg" width="70" height="70" alt="" />
+        </div>
+        {title ? <div className="qr-scanner__title">{title}</div> : null}
+        <div className="qr-scanner__error">{msg}</div>
+        <PageContainerFooter
+          onCancel={handleCancel}
+          onSubmit={tryAgain}
+          cancelText={t('cancel')}
+          submitText={t('tryAgain')}
+          submitButtonType="confirm"
+        />
+      </>
+    );
+  };
+
+  const renderVideo = () => {
+    let message;
+    if (ready === READY_STATE.ACCESSING_CAMERA) {
+      message = t('accessingYourCamera');
+    } else if (ready === READY_STATE.READY) {
+      message = t('scanInstructions');
+    } else if (ready === READY_STATE.NEED_TO_ALLOW_ACCESS) {
+      message = t('youNeedToAllowCameraAccess');
+    }
+
+    return (
+      <>
+        <div className="qr-scanner__content">
+          <div className="qr-scanner__content__video-wrapper">
+            <QrReader
+              delay={300}
+              onError={handleError}
+              onScan={handleScan}
+              style={{ width: '100%', filter: 'blur(4px)' }}
+            />
+          </div>
+        </div>
+        <div className="qr-scanner__status">{message}</div>
```

Comment:
`message` may be `undefined` here, so we may want to conditionally render this `div`.

Reply:
Cool